### PR TITLE
Add experimental Shardy support.

### PR DESCRIPTION
**Pull request by jreiffers**
This is experimental and not intended for production use. If you try this, please file issues for problems you encounter.

# Description

Shardy is a new partitioning system in JAX. It currently replaces the sharding propagation passes in GSPMD.

This PR adds Shardy support for TE's attention, normalization, quantization and softmax primitives.

To use this, simply enable Shardy:

```
jax.config.update("jax_use_shardy_partitioner", True)
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Adds functions to provide Shardy sharding rules to primitives.

### CODE REVIEW

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 1]
```diff
@@ -165,6 +210,26 @@ def get_scale_shape(
...
+
+        Args:
+            input_rank: The rank of the input tensor (for which we produce the scale tensor)
+            unique_var: An otherwise unused Shardy variable name prefix
+
+        Returns:
+            The Shardy rules for the scaling mode
+        """
+        input_spec = [f"x{i}" for i in range(input_rank)]
+        input_spec[-2] = CompoundFactor(unique_var, "block_size")
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/scaling_modes.py:224]: Hi, could you help me to understand why the `-2` axis of `input_spec` here is taken to embed with block scaling information?

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 2]
```diff
@@ -165,6 +210,26 @@ def get_scale_shape(
...
+
+        rowwise = input_spec[:-1] + [f"{unique_var}_"]
+        colwise = input_spec[:-2] + [unique_var, f"{unique_var}__"]
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/scaling_modes.py:227]: Why `rowwise` and `colwise` have different specs?

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 3]
```diff
@@ -534,6 +535,60 @@ def sharded_impl(x, scale, gamma, beta):
...
+            result_types,
+        )
+
+        scale_rules = ScalingMode(scaling_mode).get_shardy_sharding_rules(
+            len(value_types[0].shape), unique_var="i"
+        )
+        x_axes = scale_rules.input_spec
+
+        out = x_axes[:-1] + ("k",)
+        if is_2x and norm_type != NVTE_Norm_Type.RMSNorm:
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/normalization.py:570]: Why do we exclude `RMSNorm` here?

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 4]
```diff
@@ -534,6 +535,60 @@ def sharded_impl(x, scale, gamma, beta):
...
+            colwise_out = out
+        else:
+            colwise_out = ("…4",)
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/normalization.py:573]: Does it mean `colwise_out` is the same as the sharding rule of `amax`? If yes, can we explicitly specify None sharding instead?

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 5]
```diff
@@ -142,9 +142,11 @@ def layernorm_fp8_mlp_prim_func(
     @pytest_parametrize_wrapper("dtype", DTYPES)
     @pytest_parametrize_wrapper("use_bias", [True, False])
     @pytest_parametrize_wrapper("fp8_recipe", SUPPORTED_RECIPES)
+    @pytest_parametrize_wrapper("use_shardy", [True, False])
```
Comment (phu0ngng) [inline tests/jax/test_distributed_layernorm_mlp.py:145]: In general, we are trying to reduce the L0 unit test time as much as possible, and we wouldn't want the testing time to increase linearly with the number of added features.

Could we split all the shardy tests to separate tests, and instead of parameterizing all input combinations, we only test sharding with a subset of input parameters?

 Here, since different input shapes, data types, and activation types are already tested without shardy, we can test shardy with a single choice of input shape, data types, and gated activation.
 
 The same logic applies to other unit tests.

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 4 above)
Comment (jreiffers) [reply transformer_engine/jax/cpp_extensions/normalization.py:573]: This is the equivalent of `None`, because `…4` is not used otherwise.

File: transformer_engine/jax/cpp_extensions/normalization.py (hunk 3 above)
Comment (jreiffers) [reply transformer_engine/jax/cpp_extensions/normalization.py:570]: Good catch. I don't recall, might just be a bad rebase or other mess-up. I'll check.

File: transformer_engine/jax/quantize/scaling_modes.py (hunk 2 above)
Comment (jreiffers) [reply transformer_engine/jax/quantize/scaling_modes.py:227]: They have different shapes, so they require different specs in Shardy.

File: tests/jax/test_distributed_layernorm_mlp.py (hunk 5 above)
Comment (jreiffers) [reply tests/jax/test_distributed_layernorm_mlp.py:145]: Does the approach I took in `test_distributed_fused_attn` look OK to you? I didn't touch the other ones, because the number of tests didn't seem significant enough, but adding that is easy enough.

**Comment by jreiffers** [comment]
PTAL. If this looks OK in principle, I'll do the rebase.

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 6]
```diff
@@ -406,6 +407,58 @@ def sharded_impl(x, scale):
...
+        del out_dtype, act_enum, act_len, scale_dtype, scale_shapes, is_outer, mesh, result_types
+
+        x_rank = len(value_types[0].shape)
+        scale_rules = ScalingMode(scaling_mode).get_shardy_sharding_rules(
+            x_rank - 1, unique_var="i"
+        )
+        x_axes = scale_rules.input_spec + (f"x{x_rank-1}",)
+        out = (*x_axes[:-2], x_axes[-1])
+        if scaling_mode == ScalingMode.NVTE_MXFP8_1D_SCALING.value:
+            scale_inv = scale_rules.rowwise_rule[:-1] + (scale_rules.rowwise_rule[-1],)
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/activation.py:433]: Hi, 
Isn't this equivalent to `scale_inv = scale_rules.rowwise_rule`?

Anyway, for ActLuPrimitive, we hardcoded the `flatten_axis=-2`, meaning that given an input shape `[M, 2, K]`, the rowwise_scale_inv.shape will be `[M, 2, K/32]` while the colwise_scale_inv.shape will be `[M/32, 2, K]`. I guess we need to update this accordingly somehow.

File: transformer_engine/jax/quantize/scaling_modes.py [hunk 7]
```diff
@@ -217,6 +262,41 @@ def get_scale_shape(
...
+        """
+        bx = unique_var
+        by = f"{unique_var}_"
+
+        # We have to use two different factors in the two CompoundFactors because of Shardy
+        # verifier requirements, even though they are the same.
+        input_spec = [f"x{i}" for i in range(input_rank - 2)] + [
+            CompoundFactor(bx, "block_size_x"),
+            CompoundFactor(by, "block_size_y"),
+        ]
```
Comment (phu0ngng) [inline transformer_engine/jax/quantize/scaling_modes.py:283]: We should take into account the `flatten_axis` here.

We quantize the [-1] axis for rowwise and [flatten - 1] axis for colwise.
For example, given a input_shape `dim0, dim1, dim2, dim3`, then:
```
flatten_axis    rowwise_scale_shape.                     colwise_scale_shape
-1                   [dim0, dim1, dim2, dim3/32]           [dim0, dim1, dim2/32, dim3] 
-2                   [dim0, dim1, dim2, dim3/32]           [dim0, dim1/32, dim2, dim3]
```

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 8]
```diff
@@ -819,6 +872,44 @@ def sharded_impl(dz, x, scale):
...
+                colwise_out = tuple(x_axes)
+        else:
+            colwise_out = ("j",)
+
+        dbias = x_axes[-2:] if is_dbias else ("k",)
+        amax = ("…4",)
+
+        return SdyShardingRule(
+            (("…0",), tuple(x_axes), ("…2",)),
+            (out, colwise_out, scale_rules.rowwise_rule, scale_rules.colwise_rule, amax, dbias),
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/activation.py:909]: Same for `DActLuDBiasQuantizePrimitive`, we hardcoded the flatten_axis=-2 and the last dim spec may not be None. So the `colwise_scale` and `rowwise_scale` may need to be reworked here.

File: transformer_engine/jax/cpp_extensions/activation.py (hunk 6 above)
Comment (jreiffers) [reply transformer_engine/jax/cpp_extensions/activation.py:433]: Yeah, this was a bad rebase. Thanks for noticing.

> Anyway, for ActLuPrimitive, we hardcoded the flatten_axis=-2, meaning that given an input shape [M, 2, K], the rowwise_scale_inv.shape will be [M, 2, K/32] while the colwise_scale_inv.shape will be [M/32, 2, K]. I guess we need to update this accordingly somehow.

Can you elaborate?

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 9]
```diff
@@ -406,6 +407,58 @@ def sharded_impl(x, scale):
...
+            scale_inv = scale_rules.rowwise_rule[:-1] + (scale_rules.rowwise_rule[-1],)
+            colwise_scale_inv = scale_rules.colwise_rule[:-1] + (scale_rules.colwise_rule[-1],)
+        else:
+            scale_inv = scale_rules.rowwise_rule
+            colwise_scale_inv = scale_rules.colwise_rule
+
+        if is_2x:
+            if scaling_mode == ScalingMode.DELAYED_TENSOR_SCALING.value:
+                colwise_out = tuple(
+                    multidim_transpose(x_axes, static_axis_boundary=-1, transpose_axis=-1)
```
[the remark this answers is not in the document]
Comment (jreiffers) [reply transformer_engine/jax/cpp_extensions/activation.py:442]: Done.

File: transformer_engine/jax/quantize/scaling_modes.py (hunk 7 above)
Comment (jreiffers) [reply transformer_engine/jax/quantize/scaling_modes.py]: Done, thanks for the example!

File: transformer_engine/jax/cpp_extensions/activation.py [hunk 10]
```diff
@@ -406,6 +407,54 @@ def sharded_impl(x, scale):
...
+        is_outer,
+        mesh,
+        value_types,
+        result_types,
+    ):
+        del out_dtype, act_enum, act_len, scale_dtype, scale_shapes, is_outer, mesh, result_types
+
+        x_rank = len(value_types[0].shape)
+        scale_rules = ScalingMode(scaling_mode).get_shardy_sharding_rules(
+            x_rank - 1, unique_var="i", flatten_axis=-1
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/activation.py:428]:
```suggestion
            x_rank - 1, unique_var="i", flatten_axis=-2
```

File: transformer_engine/jax/cpp_extensions/normalization.py [hunk 11]
```diff
@@ -519,6 +520,57 @@ def sharded_impl(x, scale, gamma, beta):
...
+            out_dtype,
+            scale_dtype,
+            scale_shapes,
+            is_outer,
+            mesh,
+            result_types,
+        )
+
+        scale_rules = ScalingMode(scaling_mode).get_shardy_sharding_rules(
+            len(value_types[0].shape), unique_var="i"
```
Comment (phu0ngng) [inline transformer_engine/jax/cpp_extensions/normalization.py:550]:
```suggestion
            len(value_types[0].shape), unique_var="i", flatten_axis=-1
```

**Review by phu0ngng — CHANGES_REQUESTED** [review]
The rest looks good to me!

File: tests/jax/test_distributed_layernorm_mlp.py [hunk 12]
```diff
@@ -257,9 +251,59 @@ def test_layernorm_mlp_grad(
...
+        mesh_config,
+        activation_type,
+        use_bias,
+        input_shape,
+        dtype,
+        use_fp8,
+        fp8_recipe,
+        use_shardy,
     ):
+        jax.config.update("jax_use_shardy_partitioner", use_shardy)
```
Comment (phu0ngng) [inline tests/jax/test_distributed_layernorm_mlp.py:306]: Hi, 

Here, I think we need to set `jax.config.update("jax_use_shardy_partitioner", False)` after the test.
Otherwise, later tests use Shardy, as it was turned on at one point but never turned off.

File: examples/jax/encoder/test_model_parallel_encoder.py [hunk 13]
```diff
@@ -503,6 +507,13 @@ def test_te_mxfp8_with_sp(self):
         actual = train_and_evaluate(self.args)
         assert actual[0] < 0.455 and actual[1] > 0.785
 
+    @unittest.skipIf(not is_bf16_supported(), "Device compute capability 8.0+ is required for BF16")
+    def test_te_bf16_shardy(self):
+        """Test Transformer Engine with BF16"""
+        self.args.enable_shardy = True
+        actual = train_and_evaluate(self.args)
+        assert actual[0] < 0.45 and actual[1] > 0.79
```
Comment (phu0ngng) [inline examples/jax/encoder/test_model_parallel_encoder.py:515]:
```suggestion
        assert actual[0] < 0.45 and actual[1] > 0.79
```
```suggestion
        assert actual[0] < 0.455 and actual[1] > 0.785
```
So that it has the same tols as `test_te_bf16`. The test passes after relaxing the tols.

**Review by phu0ngng — APPROVED** [review]

**Comment by phu0ngng** [comment]
Pipeline #26929185.
